codalab/lib/bundle_store.py

```diff
@@ -383,3 +395,76 @@ def cleanup(self, uuid, dry_run):
         if not dry_run:
             path_util.remove(absolute_path)
 
+
+    def health_check(self, model):
+        """
+        MultiDiskBundleStore.health_check(): In the MultiDiskBundleStore, bundle contents are stored on disk, and
+        occasionally the disk gets out of sync with the database, in which case we make repairs in the following ways:
+
+            1. Creates a trash directory
+            2. Moves the contents of bundles with corresponding UUID not in the database to trash
+            3. Any files that don't begin with a UUID string should be moved to trash
+            4. For each bundle marked READY or FAILED, ensure that its dependencies are not located in the bundle
+               directory.
+            5. Bundles that are stored in <UUID>.cid or <UUID>.status/ should be marked as READY or FAILED
+            6. <UUID>.sh and <UUID>-internal.sh should be gone for bundles marked READY or FAILED
+        """
+        # Create a trash directory
+        trash_dir = os.path.join(self.codalab_home, 'trash')
+        path_util.make_directory(trash_dir)
+
+        def _trash_bundle(bundle_loc, trash_loc):
+            print >> sys.stderr, 'Moving %s to trash...' % bundle_loc
+            trash_path = os.path.join(trash_loc, os.path.basename(bundle_loc))
+            path_util.rename(bundle_loc, trash_path)
+
+        # Scan files in the data directory for every partition, moving bundles that are marked as garbage to the trash
+        partitions, _ = path_util.ls(self.partitions)
+        trash_count = 0
+        for partition in partitions:
+            partition_path = os.path.join(self.partitions, partition, MultiDiskBundleStore.DATA_SUBDIRECTORY)
+            for entry in reduce(lambda d,f: d + f, path_util.ls(partition_path)):
+                bundle = os.path.join(partition_path, entry)
+                if self.__check_should_trash(bundle, model):
+                    _trash_bundle(bundle, trash_dir)
+                    trash_count += 0
+
+        print >> sys.stderr, 'Moved %d objects to the trash at %s' % (trash_count, trash_dir)
+
+
+    def __check_should_trash(self, bundle_path, model):
+        """Checks a path to a bundle to see if it needs to be trashed.
+        """
+
+        UUID_REGEX = re.compile(r'^(0x[0-9a-z]{32})')
+        file_name = os.path.basename(bundle_path)
+        uuid_match = UUID_REGEX.match(file_name)
+        if not uuid_match:
+            return True
+
+        uuid = uuid_match.groups()[0]
+
+        # Check if bundle is in the database.
+        try:
+            bundle_db = model.get_bundle(uuid)
+        except:
+            return True
+
+        if bundle_db.state in [State.READY, State.FAILED]:
+            # This is the set of paths that should be gone if the bundle is either READY or FAILED
+            # If these paths are still around then something has gone wrong and we should move the bundle to the trash.
```

Comment:
Shouldn't delete the whole bundle!  Should just delete the dependencies inside the bundle.